% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/functions.r
\name{gene_clump}
\alias{gene_clump}
\title{Clumping of gene-based test statistics}
\usage{
gene_clump(
  gentres,
  ld_population,
  chromosome = "chr",
  gene_start = "gene_start",
  gene_symbol = "symbol",
  pval = "pval",
  clump_p = 0.05/12727,
  clump_kb = 1000,
  clump_r2 = 0.01,
  verbose = TRUE
)
}
\arguments{
\item{gentres}{Gene-based association test statistic results for multiple genes (e.g., output of <gent/mugent/xgent>_genomewide())}

\item{ld_population}{one of 'EUR', 'AFR', 'EAS', 'SAS', or 'AMR' from 1000 Genomes Phase 3. Used for internal loading of GenT correlation matrices.}

\item{chromosome}{Chromosome variable name in \code{genedf}}

\item{gene_start}{Gene start position variable name in \code{genedf}}

\item{pval}{Gene-based test statistic P-value variable name in \code{genedf}}

\item{clump_p}{Only genes with a P-value less than this threshold may index a locus for clumping}

\item{clump_kb}{Kilobase size of the entire clumping window. Left and right windows from the index gene will be half the size of \code{clump_kb}}

\item{clump_r2}{Only genes correlated with lead genes beyond this threshold may be clumped to other genes}

\item{verbose}{TRUE if progress should be printed to the console, FALSE otherwise}

\item{symbol}{Gene symbol variable name in \code{genedf}}
}
\value{
A vector of clumped genes
}
\description{
This function performs PLINK-style clumping of gene-based test statistics
}
\examples{
gent_results=gent_genomewide(gwas_data,50,'EUR','ld_directory')
gene_clump(gent_results,'EUR')
}
