% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/functions.r
\name{gent_finemap}
\alias{gent_finemap}
\title{Fine-mapping gene-based association test statistics}
\usage{
gent_finemap(
  gent_results,
  ld_population,
  gwas_n,
  index_genes = NULL,
  chromosome = "chr",
  gene_start = "gene_start",
  gene_symbol = "gene",
  pval = "pval",
  null_mean = "mu_h0",
  null_variance = "sigma2_h0",
  window_kb_width = 2000,
  R_ridge_penalty = 0,
  clump_p = 0.05/12727,
  clump_r2 = 0.01,
  verbose = TRUE,
  index = NULL,
  ...
)
}
\arguments{
\item{gent_results}{Direct output of \code{gent_genomwide()} or \code{mugent_genomewide()$MuGenT}}

\item{ld_population}{LD population, one of 'EUR', 'AFR', 'EAS', 'SAS', or 'AMR'}

\item{gwas_n}{GWAS sample size}

\item{index_genes}{Vector of index genes to perform fine-mapping around. Will be found automatically using clumping if NULL}

\item{chromosome}{Column name of gene chromosome in \code{gent_results}}

\item{gene_start}{Column name of gene start position (hg19) in \code{gent_results}}

\item{gene_symbol}{Column name of gene symbol in \code{gent_results}}

\item{pval}{Column name of gene-based test P-value in \code{gent_results}}

\item{null_mean}{Column name of gene-based test null mean (number of tested SNPs) in \code{gent_results}}

\item{null_variance}{Column name of gene-based test null variance in \code{gent_results}}

\item{window_kb_width}{Kb of window in which fine-mapping should be performed (total window will be this size)}

\item{R_ridge_penalty}{Ridge penalty to add to gene-gene correlation matrix}

\item{clump_p}{P-value threshold to use when finding loci in which to perform fine-mapping (only invoked if \code{index_genes} is NULL)}

\item{clump_r2}{Gene-gene squared correlation upper threshold to use when finding loci in which to perform fine-mapping (only envoked if \code{index_genes} is NULL). Genes correlated below the square root of this threshold will be considered as tagging separate loci.}

\item{verbose}{TRUE if progress should be printed to the console, FALSE otherwise}

\item{index}{Gene index file. see \code{data(EnsemblHg19GenePos)} for the expected format.}
}
\value{
A dataframe with these components:
\itemize{
\item \code{gene}: Gene symbol
\item \code{locus_index_gene}: This gene is at the center of the fine-mapped locus and was either the most significant gene or was set by you
\item \code{chr}: Chromosome location
\item \code{gene_start}: Start position of gene using Ensembl hg19
\item \code{finemapping_pip}: SuSiE posterior inclusion probability (PIP)
\item \code{finemapping_CS}: SuSiE credible set (NA if not in any credible set; set numbering is arbitrary)
}
}
\description{
This function performs fine-mapping of gene-based association tests statistics.
}
\examples{
# Example for Alzheimer's disease GWAS data
gent_finemap(
  gent_results=gent_results,
  ld_population='EUR',
  gwas_n=50000,
  index_genes=c('CR1','MYCL'),
  window_kb_width=2000,
  verbose=TRUE)
}
